### PR TITLE
Integrate Dagger 2

### LINKED ISSUE #3751: Dependency Injection solution
### Is your feature request related to a problem? Please describe.
This is a follow-up on #3600. After removing static preferences helpers from the code, we can start injecting it properly.

We're currently not using any fancy DI framework.

### Describe the solution you'd like
I believe it is a time to start thinking about dependency injection solution of some kind, to avoid manual deps creation everywhere.

### Describe alternatives you've considered

1. choosing existing solution (pretty much only Dagger 2, fortunately maintained by Google)
2. rolling own, simpler solution, such as a flat service registry

### PR DESCRIPTION
**Pull request by ezaquarii**
Closes #3751

Signed-off-by: Chris Narkiewicz <hello@ezaquarii.com>

### CODE REVIEW

File: src/main/java/com/owncloud/android/MainApp.java [hunk 1]
```diff
@@ -113,8 +126,21 @@
 
     private static boolean mOnlyOnDevice;
 
+    private DependencyInjector dependencyInjector;
```
[the remark this answers is not in the document]
Comment (ezaquarii) [reply src/main/java/com/owncloud/android/MainApp.java]: Perhaps, maybe not.

File: src/main/java/com/owncloud/android/MainApp.java [hunk 2]
```diff
@@ -93,7 +103,10 @@
  * Contains methods to build the "static" strings. These strings were before constants in different
  * classes
  */
-public class MainApp extends MultiDexApplication {
+public class MainApp extends MultiDexApplication implements
```
[the remark this answers is not in the document]
Comment (ezaquarii) [reply src/main/java/com/owncloud/android/MainApp.java]: False positive.

File: src/main/java/com/owncloud/android/ui/fragment/FileDetailFragment.java [hunk 3]
```diff
@@ -75,7 +78,7 @@
 /**
  * This Fragment is used to display the details about a file.
  */
-public class FileDetailFragment extends FileFragment implements OnClickListener {
+public class FileDetailFragment extends FileFragment implements OnClickListener, Injectable {
```
[the remark this answers is not in the document]
Comment (ezaquarii) [reply src/main/java/com/owncloud/android/ui/fragment/FileDetailFragment.java]: False positive.

File: src/main/java/com/owncloud/android/ui/activity/ReceiveExternalFilesActivity.java [hunk 4]
```diff
@@ -577,14 +578,9 @@ private String renameSafeFilename(String filename) {
             safeFilename = safeFilename.replaceAll("=", "_");
             safeFilename = safeFilename.replaceAll(",", "_");
 
-            try {
```
Comment (ezaquarii) [inline src/main/java/com/owncloud/android/ui/activity/ReceiveExternalFilesActivity.java]: Cargo-culting findbugs.

File: src/main/java/com/owncloud/android/ui/fragment/FileDetailFragment.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (ezaquarii) [reply src/main/java/com/owncloud/android/ui/fragment/FileDetailFragment.java]: False positive.

File: src/main/java/com/owncloud/android/MainApp.java [hunk 5]
```diff
@@ -93,7 +104,10 @@
...
```
[the remark this answers is not in the document]
Comment (ezaquarii) [reply src/main/java/com/owncloud/android/MainApp.java]: False positive.

[the remark this answers is not in the document]
Comment (ezaquarii) [reply src/main/java/com/owncloud/android/MainApp.java]: We need to ignore this - those methods I added are required by Dagger. Not much of a choice here.

File: src/main/java/com/owncloud/android/ui/fragment/FileDetailFragment.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (ezaquarii) [reply src/main/java/com/owncloud/android/ui/fragment/FileDetailFragment.java]: False positive? No new methods added - this interface is required to enable dagger injector on this type.

**Comment by AndyScherzinger** [comment]
Hi @ezaquarii is this PR ready-to-review from your point of view? Just asking since it is labelled `1. developing` while it seems this could be merged as a first step/example.

**Comment by tobiasKaminsky** [comment]
Looks awesome :+1: 
Can you change the label to "2. to review" once it is done? :heart:

**Comment by tobiasKaminsky** [comment]
https://github.com/nextcloud/android/pull/3758/files#diff-b0fb3ea404c4520c55c0a642245615dfR168

do we need this still? I do not think so. At least on a test with commenting this out, it still worked…?

File: src/main/java/com/nextcloud/client/di/ActivityInjector.java [hunk 6]
```diff
@@ -0,0 +1,75 @@
+/*
+ * Nextcloud Android client application
+ *
+ * @author Chris Narkiewicz
+ * Copyright (C) 2019 Chris Narkiewicz <hello@ezaquarii.com>
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * at your option) any later version.
```
Comment (tobiasKaminsky) [inline src/main/java/com/nextcloud/client/di/ActivityInjector.java]: There seems to be a missing ( at start.

File: src/main/java/com/nextcloud/client/di/ActivityInjector.java [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+
+import androidx.fragment.app.FragmentActivity;
+import androidx.fragment.app.FragmentManager;
+import dagger.android.AndroidInjection;
+
+public class ActivityInjector implements Application.ActivityLifecycleCallbacks {
+
+    @Override
+    public final void onActivityCreated(Activity activity, Bundle savedInstanceState) {
+        if(activity instanceof Injectable) {
```
Comment (tobiasKaminsky) [inline src/main/java/com/nextcloud/client/di/ActivityInjector.java]: There is a missing whitespace after if.

We use mainly the codestyle which is default in AndroidStudio, so you can use Code -> Reformat Code.

File: src/main/java/com/nextcloud/client/di/FragmentInjector.java [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+import android.content.Context;
+
+import androidx.annotation.NonNull;
+import androidx.fragment.app.Fragment;
+import androidx.fragment.app.FragmentManager;
+import dagger.android.support.AndroidSupportInjection;
+
+class FragmentInjector extends FragmentManager.FragmentLifecycleCallbacks {
+    @Override
+    public void onFragmentPreAttached(@NonNull FragmentManager fm, @NonNull Fragment f, @NonNull Context context) {
```
Comment (tobiasKaminsky) [inline src/main/java/com/nextcloud/client/di/FragmentInjector.java]: We also usually write the full name.

File: src/main/java/com/owncloud/android/MainApp.java [hunk 9]
```diff
@@ -124,6 +150,14 @@
     @Override
     public void onCreate() {
         super.onCreate();
+
+        DaggerAppComponent.builder()
+            .application(this)
+            .build()
+            .inject(this);
```
Comment (tobiasKaminsky) [inline src/main/java/com/owncloud/android/MainApp.java]: For my personal understanding: this is something we then need in each and very Fragment/Activity which uses Dagger? 
Along with "@Inject".

And on testing we can just mock the field, right?

Comment (ezaquarii) [reply src/main/java/com/owncloud/android/MainApp.java]: You don't have to call it in each activity. It is done automatically in `FragmentInjector` and `ActivityInjector`:

    @SuppressFBWarnings("ST")
    @Override
    public void onCreate() {
        super.onCreate();

        DaggerAppComponent.builder()
            .application(this)
            .build()
            .inject(this);

        registerActivityLifecycleCallbacks(new ActivityInjector()); // magic is done here
        ...
}
```

**Comment by ezaquarii** [comment]
> Just did a small review, but to be honest I either need to dig real deep into dagger, or just do a smoke test see_no_evil

Sure, take your time. I generally prefer to ship well-reviewed code than being sad later.

File: src/main/java/com/owncloud/android/MainApp.java (hunk 9 above)
Comment (ezaquarii) [reply src/main/java/com/owncloud/android/MainApp.java]: Sadly, this automagic does not work with `Service`s, as there is no app-wide lifecycle hooks that you can plug into. You need to call `inject()` manually in `Service`, although it's a 1-liner.

File: src/main/java/com/nextcloud/client/di/ActivityInjector.java (hunk 7 above)
Comment (ezaquarii) [reply src/main/java/com/nextcloud/client/di/ActivityInjector.java]: Fixed code formatting.

File: src/main/java/com/nextcloud/client/di/FragmentInjector.java (hunk 8 above)
Comment (ezaquarii) [reply src/main/java/com/nextcloud/client/di/FragmentInjector.java]: Fixed params naming.

File: src/main/java/com/nextcloud/client/di/ActivityInjector.java (hunk 6 above)
Comment (ezaquarii) [reply src/main/java/com/nextcloud/client/di/ActivityInjector.java]: Added missing (

File: src/main/java/com/owncloud/android/MainApp.java (hunk 9 above)
Comment (tobiasKaminsky) [reply src/main/java/com/owncloud/android/MainApp.java]: Wow. This is indeed very nice.
:+1: thanks for the work!

**Comment by tobiasKaminsky** [comment]
https://github.com/nextcloud/android/blob/5d57293574c6e3ba0e456dac0d75ee209272118d/src/main/java/com/owncloud/android/ui/fragment/FileDetailFragment.java#L223

Is this still needed as we have
 
https://github.com/nextcloud/android/blob/5d57293574c6e3ba0e456dac0d75ee209272118d/src/main/java/com/owncloud/android/ui/fragment/FileDetailFragment.java#L148-L149

**Comment by ezaquarii** [comment]
> Is this still needed as we have

No, this is not needed. Thanks for catching it. I updated the branch.

Now, if you put a breakpoint at `FileDisplayActivity.java:220`, you can observe how the `preferences` are injected when you step over `super.onCreate(savedInstanceState);`.

**Comment by AndyScherzinger** [comment]
I'd vote to get this merged for 3.6.0 RC1 @tobiasKaminsky @ezaquarii

**Comment by tobiasKaminsky** [comment]
I removed two unneeded initializations and tested it.
Works really fine.

(I have a deeper look into Dagger 2 when doing one PR that is depending on it).

**Comment by tobiasKaminsky** [comment]
Thank you for your great effort on this @ezaquarii!

**Comment by AndyScherzinger** [comment]
Sweet! 🚀 
Now we can give it a test drive out in the wild with the release candidate 🥇

**Pull request merged**